corsid/iefc.py

```diff
@@ -0,0 +1,163 @@
+    A_err = A - A_pred
+    return d.l2sqr(A_err.ravel()) / d.l2sqr(A.ravel())
+
+@dataclass
+class FitToIEFCResult:
+    G: ArrayLike
+    costs: ArrayLike
+    dz_errors: ArrayLike
+    dx_errors: ArrayLike
+    z_errors: ArrayLike
+    val_errors: ArrayLike = None
+
+
+def fit_iefc_matrix(zs, us):
+    """
+    Calculate the iEFC matrix, given a time series of pairwise-probe data vectors (zs) and their
+    associated DM updates (us).
+    """
+    num_iter = len(zs.keys())
+
+    # Each z vector is shape (num_pix, num_probe).
+    # Stack into an array with shape (num_pix, num_probe, num_iter), interpreted as a stack of
+    # (num_probe, num_iter) matrices.
+    dZ = np.stack([zs[k] - zs[k-1] for k in range(1, num_iter)], axis=-1)
+
+    # Stack the u vectors into a (num_act, num_iter) matrix
+    U = np.column_stack([us[k] for k in range(1, num_iter)])
+
+    # dZ = A*U, so A_hat = dZ * pinv(U), with dZ and A looped over pixels
+    return bl.batch_mmip(dZ, np.linalg.pinv(U))
+
+def list_to_masked(lst):
+    """ Converts a list consisting of floats and None to a masked array with None masked out """
+    return np.ma.masked_invalid(np.array(lst, dtype=np.float64))
+
+def fit_jacobian_to_iefc_matrix(
+        data: TrainingData,
+        validation_data: TrainingData,
+        G0: ArrayLike,
+        method='L-BFGS-B',
+        options: Dict = None,
+        tol=1e-3,
+        eval_metrics_every=10,
+):
+    """
+    Two-step algorithm for estimating Jacobian matrix: calculate the iEFC matrix in closed form,
+    then fit the Jacobian to the iEFC matrix via numerical optimization.
+    """
+    if options is None:
+        options = {'disp': False, 'maxcor': 10, 'maxls': 100}
+
+    costs = []
+    z_errors = []
+    dx_errors = []
+    dz_errors = []
+    val_errors = []  # Same as dz_errors, but on validation data
+
+    # Obtain iEFC matrix in closed form
+    A = fit_iefc_matrix(data.zs, data.us)
+
+    # Starting guess for optimizer, containing only the values that we are optimizing
+    starting_guess = {'G': G0.astype(np.float64)}
+    unpack = util.make_unpacker(starting_guess)
+
+    calls = 0  # Incremented every time cost function is evaluated
+
+    def cost_for_optimizer(x):
+        sol = unpack(x)
+        G = sol['G']
+        J, grads = jax.value_and_grad(cost, argnums=(0,))(
+            G, data.Psi, A)
+        gradient = np.concatenate([np.array(grad).ravel() for grad in grads])
+        costs.append(J)
+        log.info(f'J: {J:0.3e}\t ||∂g|| = {np.linalg.norm(gradient):0.3e}')
+
+        # Evaluate quality-of-fit metrics
+        # Only do this after every 10 cost function evaluations to save time
+        nonlocal calls
+
+        if calls % eval_metrics_every == 0:
+            H = 4 * bl.batch_mt(bl.batch_mmip(G, data.Psi))
+            xs = estimate_states(H, data.zs)
+
+            z_errors.append(eval_z_error(H, xs, data.zs))
+            dx_errors.append(eval_dx_error(G, xs, data.us))
```

Comment:
You're correct that the `dx_errors` are measuring `x[k] - x[k-1] - G*u[k]`, averaged over `k` and also over all pixels, and that iEFC doesn't care at all about state estimates.  Here, though, we are fitting a Jacobian to the iEFC matrix, so keeping track of the consistency of the states is useful.

However, this made me realize that there's a bug here: the pairwise estimator maps `z` to `x`, and because it's just a matrix multiplication, it maps `dz` to `dx`. Therefore, I'm actually computing this error metric wrong here.

Reply:
I'm going to rename the `xs` to `dxs`, but will punt on fixing `eval_dx_error()` because it will be affected by the pairwise refactor, so I'll just add a FIXME.

Reply:
This will all have to get refactored potentially when we do the pairwise stuff.